src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -349,17 +413,39 @@ public void onParameterChanged(LXParameter parameter) {
         sendNoteOn(0, DEVICE_ON_OFF, this.effect.enabled.isOn() ? LED_ON : LED_OFF);
       } else {
         for (int i = 0; i < this.knobs.length; ++i) {
-          if (parameter == this.knobs[i]) {
+          LXListenableNormalizedParameter knobParam = parameterForKnob(this.knobs[i]);
+          if (parameter == knobParam) {
             double normalized = (parameter instanceof CompoundParameter) ?
               ((CompoundParameter) parameter).getBaseNormalized() :
-              this.knobs[i].getNormalized();
+                    knobParam.getNormalized();
             sendControlChange(0, DEVICE_KNOB + i, (int) (normalized * 127));
             break;
           }
         }
       }
     }
 
+    // Returns what parameter should be changed when a physical knob is turned. This
+    // is trivial except for LinkedColorParameters, whose function changes depending
+    // on the surface's mode.
+    private LXListenableNormalizedParameter parameterForKnob(LXListenableParameter knob) {
+      if (knob == null || knob instanceof LXListenableNormalizedParameter) {
+        // If the knob is unused or linked to a vanilla parameter, we're done.
+        return (LXListenableNormalizedParameter) knob;
+      }
+
+      // The knob is attached to an LinkableColorParameter.
+      LinkedColorParameter lcp = (LinkedColorParameter) knob;
```

Comment:
Need an explicit check here for `knob instanceof LinkedColorParameter`

If someone passes in a basic `ColorParameter` or some other kind of `AggregateParameter` this is going to throw a `ClassCastException`

I think we should also make this work for `ColorParameter` - not just `LinkedColorParameter` - in that case it would be nice for it to do the hue/sat/brightness magic with the bank buttons, without the ability to select a swatch color.

Reply:
Done